src/rgw/rgw_sal.cc

```diff
@@ -93,213 +83,86 @@ RGWObjState::RGWObjState(const RGWObjState& rhs) : obj (rhs.obj) {
-    }
-    if (rados->init_complete(dpp) < 0) {
-      delete driver;
-      return nullptr;
-    }
-
-    lsubdout(cct, rgw, 1) << "rgw_d3n: rgw_d3n_l1_local_datacache_enabled=" <<
-      cct->_conf->rgw_d3n_l1_local_datacache_enabled << dendl;
-    lsubdout(cct, rgw, 1) << "rgw_d3n: rgw_d3n_l1_datacache_persistent_path='" <<
-      cct->_conf->rgw_d3n_l1_datacache_persistent_path << "'" << dendl;
-    lsubdout(cct, rgw, 1) << "rgw_d3n: rgw_d3n_l1_datacache_size=" <<
-      cct->_conf->rgw_d3n_l1_datacache_size << dendl;
-    lsubdout(cct, rgw, 1) << "rgw_d3n: rgw_d3n_l1_evict_cache_on_start=" <<
-      cct->_conf->rgw_d3n_l1_evict_cache_on_start << dendl;
-    lsubdout(cct, rgw, 1) << "rgw_d3n: rgw_d3n_l1_fadvise=" <<
-      cct->_conf->rgw_d3n_l1_fadvise << dendl;
-    lsubdout(cct, rgw, 1) << "rgw_d3n: rgw_d3n_l1_eviction_policy=" <<
-      cct->_conf->rgw_d3n_l1_eviction_policy << dendl;
-  }
-#ifdef WITH_RADOSGW_DBSTORE
-  else if (cfg.store_name.compare("dbstore") == 0) {
-    driver = newDBStore(cct);
-
-    if ((*(rgw::sal::DBStore*)driver).set_run_lc_thread(use_lc_thread)
-                                    .initialize(cct, dpp) < 0) {
-      delete driver;
-      return nullptr;
-    }
-  }
-#endif
-
-#ifdef WITH_RADOSGW_MOTR
-  else if (cfg.store_name.compare("motr") == 0) {
-    driver = newMotrStore(cct);
-    if (driver == nullptr) {
-      ldpp_dout(dpp, 0) << "newMotrStore() failed!" << dendl;
-      return driver;
-    }
-    ((rgw::sal::MotrStore *)driver)->init_metadata_cache(dpp, cct);
-
-    return store;
-  }
-#endif
-
-#ifdef WITH_RADOSGW_DAOS
-  else if (cfg.store_name.compare("daos") == 0) {
-    driver = newDaosStore(cct);
-    if (driver == nullptr) {
-      ldpp_dout(dpp, 0) << "newDaosStore() failed!" << dendl;
-      return driver;
-    }
-    int ret = driver->initialize(cct, dpp);
-    if (ret != 0) {
-      ldpp_dout(dpp, 20) << "ERROR: store->initialize() failed: " << ret << dendl;
-      delete driver;
-      return nullptr;
-    }
+  auto dlname = fmt::format("/usr/lib64/ceph/librgw_sal_{}.so", cfg.store_name);
+  auto altdlname = fmt::format("./lib/librgw_sal_{}.so", cfg.store_name);
+  rgw::sal::Driver *store = nullptr;
+  rgw::sal::Driver *next = nullptr;
+  void *dl = nullptr;
+  rgw::sal::Driver *(*newDriver)(const DoutPrefixProvider *, CephContext *, bool, bool, bool, bool, bool, bool, bool, bool) = nullptr;
+  dl = dlopen(dlname.c_str(), RTLD_NOW | RTLD_LOCAL | RTLD_DEEPBIND);
+  if (!dl)
+    dl = dlopen(altdlname.c_str(), RTLD_NOW | RTLD_LOCAL | RTLD_DEEPBIND);
+  if (dl) {
+    newDriver = (rgw::sal::Driver* (*)(const DoutPrefixProvider *, CephContext *, bool, bool, bool, bool, bool, bool, bool, bool))dlsym(dl, "new_Driver");
+    if (newDriver)
+      store = newDriver(dpp, cct, false, use_gc_thread, use_lc_thread, quota_threads, run_sync_thread, run_reshard_thread, use_cache, use_gc);
+    if (dlclose(dl) < 0)
+      ldpp_dout(dpp, 0) << "WARNING: dlclose() failed" << dendl;
```

Comment:
is it really safe to `dlclose()` the library immediately after creating the driver?